### DISCUSSION TITLE
Include specific folders of excluded folder

### DISCUSSION THREAD

**Discussion by dgraziotin** [borgbase/vorta#2156] [FAQ] [answered]
I defined a simple backup with source `/Users/daniel`.

I want to exclude backing up `/Users/daniel/Library` 
But I want to backup `/Users/daniel/Library/Mobile Documents/` and `/Users/daniel/Library/Application Support/`.

I thought that the following would achieve this but I am likely using it wrong. 

```
/Users/daniel/Library/
+ /Users/daniel/Library/Mobile Documents/
+ /Users/daniel/Library/Application Support
```
(I also tried `! /Users/daniel/Library/`)

**Comment by goebbe** [comment, marked as answer]
@dgraziotin 
It seems that you want to use a combination of "_excludes_" and "_includes_".
This is possible, however, you have to use `--pattern` or `--patterns-from` options, via Vorta > Schedule > Shell Commands > Extra arguments for borg create. 

You can find examples on how-to use both options here: https://github.com/borgbase/vorta/issues/1991

The following "Extra arguments for borg" should work for your case: 
`--pattern='+Users/daniel/Library/Mobile Documents/' --pattern='+re:Users/daniel/Library/Application\sSupport' --pattern='- Users/daniel/Library/**/.*'`
When using `--pattern` (as a command line option), keep in mind that matching special characters can be tricky. 

If you have many or complicated patterns, it is likely better to use `--patterns-from` to point to a text-file with the patterns: https://github.com/borgbase/vorta/issues/1991#issuecomment-2354851384

Order matters! Always _start_ with include "+" patterns, _then_ specify exclude "-" patterns.
Do not use the use "!" for excluding folders, when you want to include something from these folders. 

WARNING: `--pattern` and `--patter-from` are **borg** features that are not integrated into Vorta.

**Reply by dgraziotin** [reply]
Thank you so much. I thought that these patterns would be for Vorta (Manage Excluded Items -> Raw). I see in #1991 that they are not supported yet.

**Reply by goebbe** [reply]
Exactly! Currently, Vorta (0.10.1) uses `--exlude` and `--exclude-from` internally. So currently, "includes" are not supported via the GUI.

**Reply by dgraziotin** [reply]
Now I have to figure out how to escape the space from Vorta's GUI. It complains because of the " Documents" part in 
`--pattern=+Users/daniel/Library/Mobile Documents`. ([Default] borg.exe create: error: argument ARCHIVE: "Documents": No archive specified ).
I've tried

--pattern=+Users/daniel/Library/Mobile\ Documents
--pattern="+Users/daniel/Library/Mobile Documents"
--pattern="+Users/daniel/Library/Mobile Documents"
--pattern='+Users/daniel/Library/Mobile\ Documents'
--pattern="+Users/daniel/Library/Mobile\ Documents"

But they all fail (all quoted ones result in `A pattern/command must start with any of: -,!, +, R, r, P, p`)

Edit: also quoting after the + sign fails (+"Users...")

Edit2:, apparently, only `\s` seems to work (`--pattern=+Users/daniel/Library/Mobile\sDocuments`)

**Reply by goebbe** [reply]
The use of quotes should work just fine. (Do not use "escapes" AND quotes at the same time)
~~However, you should not put the "+" inside the quotes.~~

**Reply by dgraziotin** [reply]
As in `--pattern=+"Users/daniel/Library/Mobile Documents"`? It didn't work either (see my first Edit). Only `--pattern=+Users/daniel/Library/Mobile\sDocuments` seems to work

**Reply by goebbe** [reply]
Interesting. What platform/ OS do you use?

**Reply by dgraziotin** [reply]
macOS 15.0, Vorta (and Borg) installed via homebrew, all latest versions.

<img width="705" alt="Screenshot 2024-12-05 at 10 21 45" src="https://github.com/user-attachments/assets/70c18396-cb4a-4d83-9c30-e23f7d867178">

**Reply by goebbe** [reply]
O.k. apparently we ran into an issue/inconsistency with borgbackup, here. 
I tried to reproduce on Linux - and also had trouble matching a folder with a space in its name, using `--pattern`.
My folder was named `test folder` 
- Putting the file path into _quotation marks_ (single or double) did not work. 
- Escaping the space with `\` did not work. 
- Replacing the space with `\s` did not work either, using the standard sh: matcher. 

The following worked for me. Note, however, that I had to use the "re:" matcher for **regular expressions**, in order to replace the single space by `\s`, here: 
`--pattern='+re:home/user/test\sfolder'`

In many cases, it may be sufficient to replace the space by `?`. However, this may also match other names, since `?` matches any single letter:  `--pattern='+home/user/test?sfolder'`  matches `test folder` but would also match `testAfolder` or `test2folder` and so on.

I opened an issue for borg: https://github.com/borgbackup/borg/issues/8578

**Note**: When using `--patterns-from` with a pattern file, it is possible to match as follows:  
`+home/user/test folder`
However, this does not work using `--pattern` on the command line (shell).

**Reply by goebbe** [reply]
> Only `--pattern=+Users/daniel/Library/Mobile\sDocuments` seems to work

I just tested this on Linux: While I did _not_ get an error message when simply replacing the space by \s - the pattern does not seem to actually match the folder. 
Did you check if the files/folders have actually been included in your backup?

**Reply by dgraziotin** [reply]
Yes, apologies. I took the lack of errors as a sign that it worked. In fact, it did not. I will now test `--patterns-from` and report in the Borg issue.

**Reply by goebbe** [reply]
I opened a Vorta issue: https://github.com/borgbase/vorta/issues/2164
For Vorta versions prior to 0.10.3, as a **workaround**, the following _command-line_ options work to match names with _white space_ to include items using Vorta: 
`--pattern='+re:home/user/test\sfolder'`
or 
`--pattern='+home/user/test?folder'`

**Comment by goebbe** [comment]
In Vorta (version 0.10.2 and earlier) there is an **issue** concerning the match of items names that contain **whitespace**. 
This issue **only appears with command line options** and is already fixed in the repository.

In future version of Vorta, in "Extra arguments for borg" the standard borg syntax should work to match names with whitespace.

Example patterns, for shell command arguments, to "include" an item with name `test folder`:
```
--pattern=+subdir/test\ folder         #escaping whitespace using "\" may not work with every shell
'--pattern=+subdir/test folder'
--pattern='+subdir/test folder'        #recommended
```

**Comment by goebbe** [comment]
The issue is fixed in Vorta 0.10.3. 
For pattern matching items with whitespace, one can use Vortas "Extra arguments for borg" with standard borg syntax.

To **include** a "test item" with **whitespace** use: 
`--pattern='+path/test item'`

To **exclude** a "test item" with **whitespace** use:  
`--pattern='-path/test item'`

**Warning**: `--pattern` is a **borg** feature that is _not_ integrated into Vorta. It could be that --pattern do not work as expected in combination with the integrated --exclude rules. Do the required reading and test before you use this.
